R/deploymentTarget.R

```diff
@@ -1,93 +1,241 @@
+  if (nrow(appDeployments) > 1) {
+    cli::cli_abort(
+      c(
+        "Supplied {.arg appId} ({appId}) identifies multiple deployments.",
+        i = "Manage obsolete deployments with rsconnect::forgetDeployment()."
+      ),
+      call = error_call
     )
-  } else if (nrow(appDeployments) == 1) {
-    # If both appName and appId supplied, check that they're consistent.
-    if (!is.null(appId) && appDeployments$appId != appId) {
-      cli::cli_abort(
-        c(
-          "Supplied {.arg appId} ({appId}) does not match deployment record ({appDeployments$appId}).",
-          i = "Omit {.arg appId} to use existing for deployment for app {.str {appName}}, or",
-          i = "Omit {.arg appName} to create new deployment record."
-        ),
-        call = error_call
-      )
-    }
+  }
 
-    updateDeploymentTarget(appDeployments, appTitle, envVars)
-  } else {
-    deployment <- disambiguateDeployments(appDeployments, error_call = error_call)
-    updateDeploymentTarget(deployment, appTitle, envVars)
+  # Existing local deployment record.
+  if (nrow(appDeployments) == 1) {
+    deployment <- appDeployments[1, ]
+    deployment <- updateDeploymentTarget(deployment, appTitle, envVars)
+    return(list(
+      accountDetails = accountDetails,
+      deployment = deployment
+    ))
   }
-}
 
-deploymentTargetForApp <- function(appId,
-                                   appTitle = NULL,
-                                   account = NULL,
-                                   server = NULL) {
-  accountDetails <- findAccount(account, server)
+  # No local deployment record. Get it from the server.
   application <- getApplication(accountDetails$name, accountDetails$server, appId)
 
-  createDeploymentTarget(
+  # Note: The account+server of this deployment record may
+  # not correspond to the original content creator.
+  deployment <- createDeploymentTarget(
     appName = application$name,
     appTitle = application$title %||% appTitle,
     appId = application$id,
-    envVars = NULL,
+    envVars = envVars,
     username = application$owner_username %||% accountDetails$name,
     account = accountDetails$name,
     server = accountDetails$server
   )
+
+  return(list(
+    accountDetails = accountDetails,
+    deployment = deployment
+  ))
+}
+
+# Discover the deployment target given appName.
+#
+# When appName is provided it identifies content previously created by a locally configured account.
+#
+# The account details from the deployment record identify the final credentials we will use, as
+# account+server may not have been specified by the caller.
+deploymentTargetFromAppName <- function(
+  recordPath = ".",
+  appName = NULL,
+  appTitle = NULL,
+  envVars = NULL,
+  account = NULL,
+  server = NULL,
+  forceUpdate = FALSE,
+  error_call = caller_env()
+) {
+
+  appDeployments <- deployments(
+    appPath = recordPath,
+    nameFilter = appName,
+    accountFilter = account,
```

Comment:
Why is `account` used here? i.e. what makes this path different from `deploymentTargetFromAppId()`?

Reply:
When given `appId`, that identifies a unique remote item. On one server, there cannot be two content items having the same identifier.

When using `appId`, we search across all deployment records, even when there is no local account for the account+server referenced by the deployment record.

When `appId` does not identify existing content, deployment fails.

When given `appName`, that identifies a remote item owned by the current user. Two people can use the same name to identify different items.

When using `appName`, the search is restricted to the incoming account+server. When there is no incoming account+server, the search is restricted to deployments which have a corresponding local account.

When `appName` does not identify existing content, new content is created.

Reply:
Thanks, that's helpful! I wonder if it's worth putting in as a comment in `deploymentTarget()`?